File: Chapter 06/ch6/tip1/NativeModule/ios/HelloManager.m
//
//  HelloManager.m
//  NativeModule
//
//  Created by Stanislav Bershadskiy on 6/24/16.
//

#import "HelloManager.h"

#import "RCTBridge.h"
#import "RCTEventDispatcher.h"

@implementation HelloManager

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(greetUser:(NSString *)name isAdmin:(BOOL *)isAdmin callback:(RCTResponseSenderBlock) callback) {
  NSLog(@"User Name: %@ , Administrator: %@", name, isAdmin ? @"Yes" : @"No");
  
  NSString *greeting = [NSString stringWithFormat:@"Welcome %@, you %@ an administrator",
                        name, isAdmin ? @"are" : @"are not"];

  
  callback(@[greeting]);
}


RCT_EXPORT_METHOD(greetUserWithPromises:(NSString *)name
                  isAdmin:(BOOL *)isAdmin
                  resolver: (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  
  NSLog(@"User Name: %@ , Administrator: %@", name, isAdmin ? @"Yes" : @"No");
  
  NSString *greeting = [NSString stringWithFormat:@"Welcome %@, you %@ an administrator",
                        name, isAdmin ? @"are" : @"are not"];
  
  
  resolve(@[greeting]);
  
}

RCT_EXPORT_METHOD(greetUserWithEvent:(NSString *)name isAdmin:(BOOL *)isAdmin) {
  NSLog(@"User Name: %@ , Administrator: %@", name, isAdmin ? @"Yes" : @"No");
  
  NSString *greeting = [NSString stringWithFormat:@"Welcome %@, you %@ an administrator",
                        name, isAdmin ? @"are" : @"are not"];

  [self.bridge.eventDispatcher sendAppEventWithName:@"GreetingResponse" body:@{@"greeting":greeting}];
}

@end

File: Chapter 06/ch6/tip3/NativeModule/android/app/src/main/java/com/nativemodule/HelloManager.java
package com.nativemodule;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

/**
 * Created by stan229 on 7/6/16.
 */
public class HelloManager extends ReactContextBaseJavaModule {
    public HelloManager(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "HelloManager";
    }

    @ReactMethod
    public void greetUser(String name, Boolean isAdmin, Callback callback) {
        System.out.println("User Name: " + name + ", Administrator: " + (isAdmin ? "Yes" : "No"));

        String greeting = "Welcome " + name + ", you " + (isAdmin ? "are" : "are not") + " an administrator";

        callback.invoke(greeting);
    }

    @ReactMethod
    public void greetUserWithPromises(String name, Boolean isAdmin, Promise promise) {
        System.out.println("User Name: " + name + ", Administrator: " + (isAdmin ? "Yes" : "No"));

        String greeting = "Welcome " + name + ", you " + (isAdmin ? "are" : "are not") + " an administrator";

        promise.resolve(greeting);
    }

    @ReactMethod
    public void greetUserWithEvent(String name, Boolean isAdmin) {
        System.out.println("User Name: " + name + ", Administrator: " + (isAdmin ? "Yes" : "No"));

        String greeting = "Welcome " + name + ", you " + (isAdmin ? "are" : "are not") + " an administrator";

        WritableMap body = Arguments.createMap();
        body.putString("greeting", greeting);

        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("GreetingResponse", body);
    }
}
